mlflow/tensorflow.py

```diff
@@ -554,10 +554,12 @@ class _TF2Wrapper(object):
     ``predict(data: pandas.DataFrame) -> pandas.DataFrame``. For TensorFlow versions >= 2.0.0.
     """
 
-    def __init__(self, infer):
+    def __init__(self, model, infer):
         """
+        :param model: A Tensorflow SavedModel.
         :param infer: Tensorflow function returned by a saved model that is used for inference.
         """
+        self.model = model
```

Comment:
Could you explain why add this attribute address the issue ? the `model` attribute is not used in `predict`

Reply:
Why the existing `_TF2Wrapper` doesn't work:
The TensorFlow model is loaded inside of the `_load_pyfunc` method. After loading the model the reference count of the `loaded_model` object is set to 1. When the `_load_pyfunc` method returns the reference count is decreased to 0 and the model gets deallocated. If you then try to call the infer signature, TensorFlow won't be able to find the `loaded_model` anymore.

Why the modified `_TF2Wrapper` works:
If you instead pass the model to the `_TF2Wrapper` also this object will hold a reference to the
`loaded_model` object. That means in this case the reference count is 2. When the `_load_pyfunc` method returns, the reference count is reduced to 1 but never reaches zero because the `_TF2Wrapper` still exists. This avoids that the model gets deallocated.

Hope this answers helps. If you have more questions don't hesitate to ask me.

Reply:
Make sense.
This looks more like a TF2 bug , the returned signature should keep reference to the model, (but actually the reference count is decreased to 0 ), but the workaround here is fine.

Reply:
I think the reason is that ConcreteFunctions (the signature is such a function) only retain a weak reference to the model variables. The behavior is described here: https://www.tensorflow.org/guide/function

Reply:
ok i see. Thanks!
Could you help add test for model serving on TF2 ?

Reply:
Sure I can help. 

A unit test ist already part of this pull request:
https://github.com/mlflow/mlflow/pull/4319/files/d05ba3e41325b159945169dc6ed1a71755172f24#diff-314e8120f884dcac5855f9ce3d1094d7871d4cc07d885d41b425d1f66c8d0e72


Do you need more?

Reply:
I update your added test. and move it into tests/tensorflow/test_tensorflow2_model_export.py and add `@pytest.mark.large`

Reply:
Follow-up question: could you explain why the existing test passed ? 
https://github.com/mlflow/mlflow/blob/b454522ecc7727e6d90b59969019428ee49e6779/tests/tensorflow/test_tensorflow2_model_export.py#L596

The existing test also run under Tensorflow2 and also do similar thing, the only difference is it use 'predict' signature, but you use 'serving_default' signature.

Reply:
The test passed because the `saved_tf_iris_model` was defined using the Estimator API (`tf.estimator.DNNClassifier`). The API already existed in TensorFlow 1 and  doesn’t use `tf.functions`, that’s why it works. I don’t know the TensorFlow Roadmap but I assume that Google will deprecate the Estimator API sometime in the future. The Estimator documentation (https://www.tensorflow.org/guide/estimator) says the following: 

> TensorFlow implements several pre-made Estimators. Custom estimators are still suported, but mainly as a backwards compatibility measure. Custom estimators should not be used for new code

Moreover, the TFX documentation (https://www.tensorflow.org/tfx/guide/keras) mentions:

> The Estimator API has been retained in TensorFlow 2.x, but is not the focus of new features and development.

Regarding `pyfunc_wrapper.predict`: At runtime `pyfunc_wrapper.predict` calls `_TF2Wrapper.predict` and this method also uses the signature (`self.infer`).

Reply:
@saschaschramm Thanks for the explanation ! I add related comments into code.